### ISSUE TITLE
[🐛 Bug]: "Invalid Host Header" in Firefox 106 w/ geckodriver 0.32.0

### ISSUE THREAD

**Issue by zireael-0** [SeleniumHQ/selenium#11264]
### What happened?

I'm using Firefox 106 with `selenium-firefox-driver:4.5.0` for Java and geckodriver 0.32.0 (via Selenide 6.9.0)

Selenium `GeckoDriverService` launches `geckodriver` and when starting a new browser session it calls `PUT http://localhost:<port>/session`

However, since the `--allow-hosts localhost` flag is **not** provided when starting `geckodriver`, requests to geckodriver  always return `500 Invalid Host Header`

Source of `geckodriver` arguments:

https://github.com/SeleniumHQ/selenium/blob/7fdaf217b8e58033b7cd01d916e3063c8198a403/java/src/org/openqa/selenium/firefox/GeckoDriverService.java#L187-L211

Source of target url for reqeusts to `geckodriver`:

https://github.com/SeleniumHQ/selenium/blob/7fdaf217b8e58033b7cd01d916e3063c8198a403/java/src/org/openqa/selenium/remote/service/DriverService.java#L172-L174

---

My suggestion would be to update `GeckoDriverService` to either include `--allow-hosts localhost` or override `::getPort(int)` to  `http://127.0.0.1:%d`.

### How can we reproduce the issue?

Breaks 100% of the time with the specified versions:
```java
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxOptions;

public class Main {
    public static void main(String[] args) throws Exception {
        var options = new FirefoxOptions();
        System.setProperty("webdriver.gecko.driver", "/usr/bin/geckodriver"); // geckodriver 0.32.0
        options.setBinary("/usr/bin/firefox"); // firefox 106 installed

        var driver = new FirefoxDriver(options);
        driver.get("https://www.stackoverflow.com");
    }
}
```

### Relevant log output

```shell
1668526366108	geckodriver	INFO	Listening on 127.0.0.1:55726
1668526366212	webdriver::server	WARN	Rejected request with Host header localhost:55726, allowed values are []
Exception in thread "main" org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: Invalid Host header localhost:55726
```

### Operating System

linux

### Selenium version

java 4.5.0

### What are the browser(s) and version(s) where you see this issue?

Firefox 106

### What are the browser driver(s) and version(s) where you see this issue?

geckodriver 0.32.0

**Comment by zireael-0** [comment]
After updating my `/etc/hosts` file to include (it was empty before)
```
127.0.0.1    localhost
```

The issue disappeared. Seems like `geckodriver` performs a reverse-lookup of the address?

However, I still think this should be fixed in Selenium, it's just less severe for me now.

**Comment by diemol** [comment]
@whimboo what do you think?

**Comment by whimboo** [comment]
Yes, for security reasons geckodriver has to check if the connection request comes from a local host by default. If no host is set at the system level it won't work. An alternative could be to explicitly specify hosts via `--allow-hosts` but IMHO the system configuration should be correct.

**Comment by titusfortner** [comment]
It doesn't appear to me like we allow args to be modified in the Java services instances; @diemol am I missing something? This implementation of builder pattern is confusing to me, regardless. :)

**Comment by diemol** [comment]
In theory you can add it through `addArguments` in `FirefoxOptions`. 

Closing this as it was due to a system mis-configuration.

**Comment by titusfortner** [comment]
@diemol this is arguments used to start the driver, not arguments passed in Options to the driver to start the browser.

**Comment by diemol** [comment]
Even if that is the case, isn't that a different issue? Main case here was a system misconfig.

**Comment by titusfortner** [comment]
I think it's a bug that Java doesn't allow the user to easily specify arguments sent to the driver. We can track it separately if you want.

**Comment by zireael-0** [comment]
@diemol To be honest, I don't think this is a misconfig. Yes, I was able to work around the issue by updating my system configuration. However, if you ask me, having an empty `/etc/hosts` file is not a misconfiguration - it's even the default on my distro.

I don't know if it's necessary to change this within selenium per-se. But in my opinion there 
should at least be a way to pass additional parameters to `geckodriver` for cases like this. Right now, I can set parameters to launch `firefox` via `FirefoxOptions`, but not `geckodrvier` itself.

If I had no su access to the machine I am working on, I wouldn't have been able to resolve this.

**Comment by titusfortner** [comment]
Well, you should be able to create your own `GeckoDriverService` with the arguments specified, but I haven't tested it to see if they get overridden somewhere (FirefoxService code is a bit weird for legacy reasons). The code itself encourages using the Builder, though, and that doesn't support adding arguments. I want us to do a review of it all soon.

**Comment by diemol** [comment]
I've modified the builder so the `--allow-hosts` is possible now. I was going to add all driver parameters, but then the structure of `FirefoxDriverService` - `GeckoDriverService` stopped me before making more changes. So we need to figure out that part first before moving forward.

For now, this issue is solved and it should be part of 4.9.0, which should come out in a few days.
